tests/Microsoft.Identity.Test.Unit/TelemetryTests/OTelInstrumentationTests.cs

```diff
@@ -60,7 +65,247 @@ public async Task AcquireTokenOTelTestAsync()
+
+                Trace.WriteLine("1. Setup an app with a token cache with one AT");
+
+                var miBuilder = ManagedIdentityApplicationBuilder.Create(ManagedIdentityId.SystemAssigned)
+                    .WithHttpManager(httpManager);
+
+                // Disabling shared cache options to avoid cross test pollution.
+                miBuilder.Config.AccessorOptions = null;
+
+                var mi = miBuilder.BuildConcrete();
+
+                httpManager.AddManagedIdentityMockHandler(
+                        appServiceEndpoint,
+                        resource,
+                        MockHelpers.GetMsiSuccessfulResponse(),
+                        ManagedIdentitySource.AppService);
+
+                AuthenticationResult result = await mi.AcquireTokenForManagedIdentity(resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Trace.WriteLine("2. Configure AT so that it shows it needs to be refreshed");
+                var refreshOn = TestCommon.UpdateATWithRefreshOn(mi.AppTokenCacheInternal.Accessor).RefreshOn;
+
+                Trace.WriteLine("3. Configure MSI to respond with a valid token");
+                httpManager.AddManagedIdentityMockHandler(
+                        appServiceEndpoint,
+                        resource,
+                        MockHelpers.GetMsiSuccessfulResponse(),
+                        ManagedIdentitySource.AppService);
+
+                // Act
+                Trace.WriteLine("4. ATM - should perform an RT refresh");
+                result = await mi.AcquireTokenForManagedIdentity(resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                // Assert
+                TestCommon.YieldTillSatisfied(() => httpManager.QueueSize == 0);
+
+                Assert.IsNotNull(result);
+
+                Assert.AreEqual(0, httpManager.QueueSize,
+                    "MSAL should have refreshed the token because the original AT was marked for refresh");
+                
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+                Assert.AreEqual(CacheRefreshReason.ProactivelyRefreshed, result.AuthenticationResultMetadata.CacheRefreshReason);
+                Assert.AreEqual(refreshOn, result.AuthenticationResultMetadata.RefreshOn);
+
+                result = await mi.AcquireTokenForManagedIdentity(resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+                Assert.AreEqual(CacheRefreshReason.NotApplicable, result.AuthenticationResultMetadata.CacheRefreshReason);
+
+                s_meterProvider.ForceFlush();
+                VerifyMetrics(4, _exportedMetrics, 4, 0);
+            }
+        }
+
+        private MockHttpMessageHandler AddMockHandlerAadSuccess(
```

Comment:
Duplicate code? I think we already have this in some other test helper. HttpManager maybe.